synapse/handlers/message.py

```diff
@@ -901,6 +914,30 @@ def persist_and_notify_client_event(
                         "Room alias %s does not point to the room" % (room_alias_str,),
                     )
 
+            # Check that alt_aliases is the proper form.
+            alt_aliases = event.content.get("alt_aliases", [])
+            if not isinstance(alt_aliases, (list, tuple)):
+                raise SynapseError(400, "The alt_aliases property must be a list.")
+
+            # If the old version of alt_aliases is of an unknown form,
+            # completely replace it.
+            if not isinstance(original_alt_aliases, (list, tuple)):
+                original_alt_aliases = []
+
+            # Check that each alias is currently valid.
+            new_alt_aliases = set(alt_aliases) - set(original_alt_aliases)
+            if new_alt_aliases:
+                for alias_str in new_alt_aliases:
+                    room_alias = RoomAlias.from_string(alias_str)
+                    mapping = yield directory_handler.get_association(room_alias)
+
+                    if mapping["room_id"] != event.room_id:
+                        raise SynapseError(
+                            400,
+                            "Room alias %s does not point to the room"
+                            % (room_alias_str,),
+                        )
```

Comment:
again, we should set an error code here (and do the same for the regular alias). 

We can take the opportunity to spec a new code, as per https://github.com/matrix-org/matrix-doc/pull/2432/files/8b9ea10e11bd0a460211e1826543fe17f2667e19#diff-a9ccbaea3cb53ab4c6147d0a7e7f10c2.

Reply:
Seems like `M_NOT_FOUND` would work well here, do you think we need something else? (Maybe one that says "the alias was found, but the underlying resource doesn't exist")

Reply:
I think we need something else: have taken it to the MSC as it's a spec issue.

Reply:
I've updated this based on the (current) conversation in the spec issue.